compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETransferStridedAccessPattern.cpp

```diff
@@ -0,0 +1,377 @@
+  return results;
+};
+
+/// Utility to check if any dimension from the L3 dma addressing can be combined
+/// with the innermost dimension, if so return the position of the dimension.
+/// Two dimensions (i and innermost) can be combined if the following conditions
+/// are satisfied: 1) stride[i] = innermost_stride * innermost_size;
+/// 2) offset[i] = 0.
+static FailureOr<size_t> isL3AddressingCombinable(
+    SmallVector<OpFoldResult> &dmaOffsets, SmallVector<OpFoldResult> &dmaSizes,
+    SmallVector<OpFoldResult> &dmaStrides) {
+  // Offsets could be dynamic but sizes and strides should be static.
+  std::optional<SmallVector<int64_t>> maybeSizes =
+      getConstantIntValues(dmaSizes);
+  std::optional<SmallVector<int64_t>> maybeStrides =
+      getConstantIntValues(dmaStrides);
+  if (!maybeSizes.has_value() || !maybeSizes.has_value()) {
+    return failure();
+  }
+  SmallVector<int64_t> sizeVals = maybeSizes.value();
+  SmallVector<int64_t> strideVals = maybeStrides.value();
+
+  // Get the index of the dim that can be potentially combined with the
+  // innermost dim. If there is no such dim, return the last index of the
+  // vector.
+  auto getPos = [&](SmallVector<int64_t> values, int64_t target) {
+    size_t i = 0;
+    for (; i < values.size() - 1; i++) {
+      if (values[i] == target) return i;
+    }
+    return i;
+  };
+
+  int64_t innerDimTotal = strideVals.back() * sizeVals.back();
+  size_t dimForCombine = getPos(strideVals, innerDimTotal);
+  if (dimForCombine >= (dmaSizes.size() - 1)) return failure();
+
+  std::optional<int64_t> offsetAtPos =
+      getConstantIntValue(dmaOffsets[dimForCombine]);
+  if (!offsetAtPos.has_value() || offsetAtPos.value() != 0) return failure();
+  return dimForCombine;
+}
+
+/// Utility to check if L2 dma addressing is linear. Note here the assumption is
+/// the dma ops are already canonicalized, so that the L2 addressing should be
+/// empty or 1-d vectors.
+static bool isL2AddressingLinear(SmallVector<OpFoldResult> &dmaOffsets,
+                                 SmallVector<OpFoldResult> &dmaSizes,
+                                 SmallVector<OpFoldResult> &dmaStrides) {
+  assert(dmaOffsets.size() == dmaSizes.size() &&
+         dmaOffsets.size() == dmaStrides.size() &&
+         "expected same number of source offsets and sizes");
+  if (dmaOffsets.size() == 0) return true;
+  if (dmaOffsets.size() != 1) return false;
+  if (!isConstantIntValue(dmaOffsets[0], 0)) return false;
+  if (!isConstantIntValue(dmaStrides[0], 1)) return false;
+  return true;
+}
+
+/// Utility to check if all users of the connection op statisfy the conditions
+/// for dma access pattern transfer.
+static FailureOr<bool> checkConnectionUsers(AMDAIE::ConnectionOp connectionOp) {
+  for (Operation *user : connectionOp->getUsers()) {
+    // Check if L3 addressing is combinable.
+    if (auto dmaOp = dyn_cast<AMDAIE::NpuDmaCpyNdOp>(user)) {
+      if (dmaOp.hasSourceAddressing() && dmaOp.hasTargetAddressing()) {
```

Comment:
I would be more future proof to check for the memory space on source/target instead of the addressing, as we might have both source and target addressing in the future and if you change the checks here to work with memory space, that future change will not impact this pass.